### ISSUE TITLE
Cannot use arithmetic between two `case()` function in Custom Expression

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#15107]
**Describe the bug**
Cannot use arithmetic between two case-functions like `case(...) + case(...)` in Custom Column.

**To Reproduce**
1. Custom question > Sample Dataset > Products
2. Column Column `case([Category] = "Widget", 1, 0) + case([Rating] > 4, 1, 0)`
![image](https://user-images.githubusercontent.com/1447303/110521288-d9975c00-810f-11eb-88ba-28c5b3404b29.png)
3. Click Visualize - query fails with `... Must be a valid instance of one of these clauses ...`
<details><summary>Full stacktrace</summary>

```
2021-03-09 19:41:31,840 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 4,
 :started_at #t "2021-03-09T19:41:31.140835+01:00[Europe/Copenhagen]",
 :error_type :invalid-query,
 :json_query
 {:type "query",
  :query
  {:source-table 10,
   :expressions
   {:Test
    ["+"
     ["case" [[["=" ["field-id" 107] "Widget"] 1]] {:default 0}]
     ["case" [[[">" ["field-id" 103] 4] 1]] {:default 0}]]}},
  :database 4,
  :parameters [],
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true}},
 :native nil,
 :status :failed,
 :class clojure.lang.ExceptionInfo,
 :stacktrace
 ["--> util.schema$schema_core_validator$fn__16348.invoke(schema.clj:29)"
  "query_processor.middleware.validate$validate_query$fn__48157.invoke(validate.clj:9)"
  "query_processor.middleware.normalize_query$normalize$fn__46427.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__44082.invoke(add_rows_truncated.clj:35)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__48071.invoke(results_metadata.clj:147)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__45682.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__47173.invoke(process_userland_query.clj:135)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__45625.invoke(catch_exceptions.clj:173)"
  "query_processor.reducible$async_qp$qp_STAR___37428$thunk__37429.invoke(reducible.clj:103)"
  "query_processor.reducible$async_qp$qp_STAR___37428.invoke(reducible.clj:109)"
  "query_processor.reducible$sync_qp$qp_STAR___37437$fn__37440.invoke(reducible.clj:135)"
  "query_processor.reducible$sync_qp$qp_STAR___37437.invoke(reducible.clj:134)"
  "query_processor$process_userland_query.invokeStatic(query_processor.clj:237)"
  "query_processor$process_userland_query.doInvoke(query_processor.clj:233)"
  "query_processor$fn__48203$process_query_and_save_execution_BANG___48212$fn__48215.invoke(query_processor.clj:249)"
  "query_processor$fn__48203$process_query_and_save_execution_BANG___48212.invoke(query_processor.clj:241)"
  "query_processor$fn__48247$process_query_and_save_with_max_results_constraints_BANG___48256$fn__48259.invoke(query_processor.clj:261)"
  "query_processor$fn__48247$process_query_and_save_with_max_results_constraints_BANG___48256.invoke(query_processor.clj:254)"
  "api.dataset$run_query_async$fn__54361.invoke(dataset.clj:56)"
  "query_processor.streaming$streaming_response_STAR_$fn__54340$fn__54341.invoke(streaming.clj:72)"
  "query_processor.streaming$streaming_response_STAR_$fn__54340.invoke(streaming.clj:71)"
  "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:65)"
  "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:63)"
  "async.streaming_response$do_f_async$fn__16055.invoke(streaming_response.clj:84)"],
 :context :ad-hoc,
 :error
 "Value does not match schema: {:query {:expressions {:Test (named [nil (named (named (not (some-matching-condition? a-clojure.lang.PersistentVector)) \"Must be a valid instance of one of these clauses: :field-id, :field-literal, :joined-field, :fk->, :datetime-field, :expression, :binning-strategy\") \"x\") (named (named (not (some-matching-condition? a-clojure.lang.PersistentVector)) \"Must be a valid instance of one of these clauses: :field-id, :field-literal, :joined-field, :fk->, :datetime-field, :expression, :binning-strategy\") \"y\")] \"Must be a valid instance of one of these clauses: :+, :-, :/, :*, :coalesce, :length, :floor, :ceil, :round, :abs, :power, :sqrt, :exp, :log\")}}}",
 :row_count 0,
 :running_time 0,
 :preprocessed nil,
 :ex-data
 {:type :schema.core/error,
  :value
  {:type :query,
   :query
   {:source-table 10,
    :expressions
    {:Test
     [:+ [:case [[[:= [:field-id 107] "Widget"] 1]] {:default 0}] [:case [[[:> [:field-id 103] 4] 1]] {:default 0}]]}},
   :database 4,
   :middleware {:js-int-to-string? true, :add-default-userland-constraints? true},
   :info
   {:executed-by 1,
    :context :ad-hoc,
    :nested? false,
    :query-hash
    [-5, 36, -23, -4, 33, -59, -67, 44, -97, 113, -94, 83, -85, -65, -74, 116, 17, -118, 14, -11, 11, -7, -1, -29, -28,
     91, -55, -48, 38, 14, -110, 68]},
   :constraints {:max-results 10000, :max-results-bare-rows 2000}},
  :error
  {:query
   {:expressions
    {:Test
     (named [nil (named (named (not (some-matching-condition? a-clojure.lang.PersistentVector)) "Must be a valid instance of one of these clauses: :field-id, :field-literal, :joined-field, :fk->, :datetime-field, :expression, :binning-strategy") "x") (named (named (not (some-matching-condition? a-clojure.lang.PersistentVector)) "Must be a valid instance of one of these clauses: :field-id, :field-literal, :joined-field, :fk->, :datetime-field, :expression, :binning-strategy") "y")] "Must be a valid instance of one of these clauses: :+, :-, :/, :*, :coalesce, :length, :floor, :ceil, :round, :abs, :power, :sqrt, :exp, :log")}}}},
 :data {:rows [], :cols []}}
```

</details>

**Information about your Metabase Installation:**
Tested 0.38.1 and latest `master`, which includes #14214

**Additional context**
Related to #12938 and #13122

**Comment by camsaul** [comment]
Hey @howonlee, sorry to step on your toes. I was showing @nemanjaglumac how to write backend MBQL tests and ended up fixing this in #15126

**Comment by camsaul** [comment]
Fixed by #15126
